internal/plugin/connectors/tcp/mssql/types/types.go

```diff
@@ -0,0 +1,57 @@
+package types
+
+import (
+	"context"
+	"io"
+	"net"
+)
+
+// NewMSSQLConnectorFunc represents the constructor of an mssqlConnector. It
+// exists so that its production implementation (mssql.NewConnector) can be
+// swapped out in unit tests.  Note we keep MSSQL in the name to prevent
+// confusion with the Secretless Connector.
+type NewMSSQLConnectorFunc func(dsn string) (MSSQLConnector, error)
+
+
+// MSSQLConnector captures the part of the 3rd party driver's mssql.Connector
+// type that we care about -- its "Connect" method -- in an interface.  This
+// allows us to mock that in our unit tests.
+type MSSQLConnector interface {
+	Connect(context.Context) (NetConner, error)
+}
+
+
+// NetConner is anything with a NetConn() method.  Ie, anything that can provide
+// a net.Conn.  Note this rather silly name conforms to Go standard conventions
+// for naming single method interfaces.
+type NetConner interface {
+	NetConn() net.Conn
+}
+
+// MSSQLConnectorFunc lets us treat a pure function (that matches the "Connect"
+// signature) as an MSSQLConnector interface.
+type MSSQLConnectorFunc func(context.Context) (NetConner, error)
+
+// Connect implements the MSSQLConnector interface on a pure function.
+func (fn MSSQLConnectorFunc) Connect(ctx context.Context) (NetConner, error) {
+	return fn(ctx)
```

Comment:
Yeah, the word "pure" here is misleading, especially since the phrase "pure function" is generally used to make a distinction around mutation / side effects.  I think I should delete that word.

> functions names can be named clearly and flexibly, they don't have to be named Connect()

Actually, the concern is more purely structural.  That is, to fulfill an interface you must define a method on some type.   Typically that type is a struct, which holds some data.  But what if there is not data to hold?  You could define a type as `struct{}` and then a method on that, but that's rather verbose boilerplate because really you only care about the method.  This trick let's you treat the plain go `func` as a type which fulfills the interface.  

This comes up a lot, and you'll see this trick other places in our code base as well as in the std lib, eg here:

https://golang.org/src/net/http/server.go?s=61509:61556#L1993